src/main/java/no/nav/pensjon/selvbetjeningopptjening/security/filter/RequestBasedBrukerbytte.java

```diff
@@ -0,0 +1,148 @@
+
+        switch (userType) {
+            case EXTERNAL:
+                return ingressTokenInfo.getUserId();
+            case INTERNAL:
+                return getVirtualLoggedInUserPid(session);
+            default:
+                log.error("Unexpected user type: {}", userType);
+                return null;
+        }
+    }
+
+    private RequestContext getUserContextForCheckingPermission(HttpSession session,
+                                                            TokenInfo ingressTokenInfo,
+                                                            EgressTokenSupplier egressTokenSupplier) {
+        UserType userType = ingressTokenInfo.getUserType();
+
+        switch (userType) {
+            case EXTERNAL:
+                return RequestContext.forExternalUser(ingressTokenInfo, egressTokenSupplier);
+            case INTERNAL:
+                return RequestContext.forInternalUser(ingressTokenInfo, getVirtualLoggedInUserPid(session), egressTokenSupplier);
+            default:
+                log.error("Unexpected user type: {}", userType);
+                return null;
+        }
+    }
+
+    private void byttBruker(String fullmektigPid,
+                            HttpServletRequest request,
+                            HttpServletResponse response) throws IOException {
+        log.debug("Received request for brukerbytte");
+        String onBehalfOfPid = getOnBehalfOfPid(request.getReader());
+        log.info("onBehalfOfPid: "+ onBehalfOfPid+" fullmektigPid: "+fullmektigPid);
+        if (onBehalfOfPid.equals(fullmektigPid)) {
+            log.info("Request for tilbakebrukerbytte accepted");
+            request.getSession().removeAttribute(FULLMAKT.name());
+            cookieSetter.unsetCookie(response, CookieType.ON_BEHALF_OF_PID);
+            respondBrukerbytteOk(response);
+            return;
+        }
+
+        boolean mayActOnBehalf = fullmaktFacade.mayActOnBehalfOf(onBehalfOfPid, fullmektigPid);
+
+        if (mayActOnBehalf) {
+            log.info("Request for brukerbytte accepted");
+            request.getSession().setAttribute(FULLMAKT.name(), onBehalfOfPid);
+            cookieSetter.setCookies(response, List.of(new CookieSpec(CookieType.ON_BEHALF_OF_PID, onBehalfOfPid)));
+            respondBrukerbytteOk(response);
+            auditor.auditFullmakt(fullmektigPid, onBehalfOfPid);
+            return;
+        }
+
+        respondWithError(HttpStatus.FORBIDDEN, "ingen fullmakt", response);
+    }
+
+    protected String getJson(Reader reader) throws IOException {
+        return IOUtils.toString(reader);
+    }
+
+    private String getOnBehalfOfPid(Reader reader) throws IOException {
+        String json = getJson(reader);
+        return new ObjectMapper().readValue(json, ByttBrukerRequest.class).fullmaktsgiverPid;
+    }
+
+    private static String getVirtualLoggedInUserPid(HttpSession session) {
+        return (String) session.getAttribute(VIRTUAL_LOGGED_IN_PID.name());
```

Comment:
Dette attributtet blir vel aldri satt, så tviler på at dette vil virke.
Vi har jo ingen database i opptjening-backend hvor sesjonsdata lagres.
Tipper du kan kopiere fra `getVirtualLoggedInPid` i [CookieBasedBrukerbytte](https://github.com/navikt/pensjon-selvbetjening-opptjening-backend/blob/master/src/main/java/no/nav/pensjon/selvbetjeningopptjening/security/filter/CookieBasedBrukerbytte.java#L95). Vi er da avhengig av at "bytt bruker"-requesten har `fnr` som query-param.

Reply:
Her har jeg nok ikke testet med intern bruker, for med ekstern bruker fungerer det. Har nok i for stor grad tatt det jeg gjorde før ferien for god fisk. Men uansett, hvorfor må fnr være query param? Bytt bruker er jo en post som poster både fullmaktgiverPid og innlogget pid, kan den hentes fra body i stedet?

Reply:
Sånn, nå tror jeg det har blitt mer riktig. Henter nå virtualLoggedInUserPid ved å slå opp på fullmektigPid i request body. La også til tilpasset versjon av testene fra søknaden.

Reply:
Oppdaget noen bugs i etterkant som gjorde at jeg måtte gjøre dette over flere commits, så jeg tror jeg bare markerer dette som et nytt review, så kan du fokusere på de siste commitsene som kom etter forrige review.